### PR TITLE
#10267 Continue button MUST give meaningful feedback during transitions

### PR DESCRIPTION
**Pull request by vbahinwillit**
## Description
Add an SR-only span next to the spinner icon, make the icon invisible to screen readers, and wrap them in a React.Fragment to maintain the current API and introduce a new prop for loading text safely.

## Testing done
Unit

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/applications/vaos/components/cancel/CancelAppointmentConfirmationModal.jsx [hunk 1]
```diff
@@ -22,6 +22,7 @@ export default function CancelAppointmentConfirmationModal({
       <p className="vads-u-margin-top--2">
         <LoadingButton
           isLoading={status === FETCH_STATUS.loading}
+          loadingText="Page change in progress"
```
Comment (jbalboni) [inline src/applications/vaos/components/cancel/CancelAppointmentConfirmationModal.jsx:25]: If this is the only text we're going to use, we could put it in FormButtons.jsx.

File: src/platform/site-wide/loading-button/LoadingButton.jsx [hunk 2]
```diff
@@ -2,16 +2,30 @@ import React from 'react';
...
   disabled,
   ...props
 }) {
-  const contents = isLoading ? (
-    <i className="fa fa-spinner fa-spin" />
-  ) : (
-    children
-  );
+  let contents;
+  if (loadingText) {
```
Comment (jbalboni) [inline src/platform/site-wide/loading-button/LoadingButton.jsx:12]: This looks good to me as a way of not breaking other components.

Something that came to mind just now was maybe having default text (like "loading" or something) that we can display when someone doesn't pass in custom `loadingText`. @1Copenut, what do you think?

File: src/applications/vaos/components/cancel/CancelAppointmentConfirmationModal.jsx (hunk 1 above)
Comment (vbahinwillit) [reply src/applications/vaos/components/cancel/CancelAppointmentConfirmationModal.jsx:25]: I had it like that at 1st but thought FormButton seemed like a generic component. I can change though.

Comment (jbalboni) [reply src/applications/vaos/components/cancel/CancelAppointmentConfirmationModal.jsx:25]: Yeah, I think they're generic to the form pages, so "Page change in progress" is probably ok. Though we have some scenarios where those buttons are set to loading before a page change, so we'd still want the ability to customize it.

File: src/platform/site-wide/loading-button/LoadingButton.jsx (hunk 2 above)
Comment (vbahinwillit) [reply src/platform/site-wide/loading-button/LoadingButton.jsx:12]: @1Copenut - Hi Trevor. Did you get a chance to look at this? Thanks

File: src/platform/site-wide/loading-button/LoadingButton.jsx [hunk 3]
```diff
@@ -2,16 +2,32 @@ import React from 'react';
...
+    contents = isLoading ? (
+      <>
+        <i
```
[the remark this answers is not in the document]
Comment (bkjohnson) [reply src/platform/site-wide/loading-button/LoadingButton.jsx:15]: At a glance it looks like this icon meets the expectations for [FontAwesome accessibility](https://fontawesome.com/v4.7.0/accessibility/)

File: src/platform/site-wide/loading-button/LoadingButton.jsx [hunk 4]
```diff
@@ -2,16 +2,32 @@ import React from 'react';
...
+        />
+        <span className="sr-only" role="status">
+          {loadingText}
+        </span>
+      </>
+    ) : (
+      children
+    );
+  } else {
+    contents = isLoading ? <i className="fa fa-spinner fa-spin" /> : children;
```
[the remark this answers is not in the document]
Comment (bkjohnson) [reply src/platform/site-wide/loading-button/LoadingButton.jsx:28]: This icon doesn't have `aria-hidden="true"` which could be problematic.

**Review by bkjohnson — COMMENTED** [review]
Took a :eyes: at the icons that the bot pointed out.  Tagging @1Copenut for visibility.

File: src/platform/site-wide/loading-button/LoadingButton.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (vbahinwillit) [reply src/platform/site-wide/loading-button/LoadingButton.jsx:28]: Thanks for the review. I added aria-hidden to the icon.

File: src/applications/vaos/containers/ReviewPage.jsx [hunk 5]
```diff
@@ -74,6 +74,7 @@ export class ReviewPage extends React.Component {
               submitStatus === FETCH_STATUS.failed
             }
             isLoading={submitStatus === FETCH_STATUS.loading}
+            loadingText="Page change in progress"
```
Comment (jbalboni) [inline src/applications/vaos/containers/ReviewPage.jsx:77]: This one should be something like "Submission in progress"

File: src/applications/vaos/containers/VAFacilityPage.jsx [hunk 6]
```diff
@@ -229,6 +232,7 @@ export class VAFacilityPage extends React.Component {
             continueLabel=""
             disabled={disableSubmitButton}
             pageChangeInProgress={loadingEligibility || pageChangeInProgress}
+            loadingText="Page change in progress"
```
Comment (jbalboni) [inline src/applications/vaos/containers/VAFacilityPage.jsx:235]: I think we need to do something like this here:

`loadingText={loadingEligibility ? "Checking eligibility requirements" : "Page change in progress"}`

File: src/applications/vaos/components/cancel/CancelAppointmentConfirmationModal.jsx (hunk 1 above)
Comment (jbalboni) [inline src/applications/vaos/components/cancel/CancelAppointmentConfirmationModal.jsx:25]: This one should say "Canceling appointment"

**Review by jbalboni — APPROVED** [review]
This looks good to me. I made a few text suggestions. Once you fix those, make sure to mark this as ready for review so that the platform team can approve it.

File: src/applications/vaos/containers/ReviewPage.jsx (hunk 5 above)
Comment (vbahinwillit) [reply src/applications/vaos/containers/ReviewPage.jsx:77]: Done.

File: src/applications/vaos/containers/VAFacilityPage.jsx (hunk 6 above)
Comment (vbahinwillit) [reply src/applications/vaos/containers/VAFacilityPage.jsx:235]: Done.

File: src/platform/site-wide/loading-button/LoadingButton.jsx [hunk 7]
```diff
@@ -2,16 +2,40 @@ import React from 'react';
...
+    contents = isLoading ? (
+      <i
+        className="fa fa-spinner fa-spin"
+        aria-hidden="true"
+        role="presentation"
+      />
+    ) : (
+      children
+    );
+  }
```
Comment (bkjohnson) [inline src/platform/site-wide/loading-button/LoadingButton.jsx:37]: Would this be possible instead?  I feel like it is a bit simpler:

```js
  const contents = isLoading ? (
    <>
      <i
        className="fa fa-spinner fa-spin"
        aria-hidden="true"
        role="presentation"
      />
      {loadingText && (
        <span className="sr-only" role="status">
          {loadingText}
        </span>
      )}
    </>
  ) : (
    children
  );
```

**Review by bkjohnson — COMMENTED** [review]
Just a comment on potential refactoring.

@ncksllvn or @kelsonic , do you have any thoughts on the site-wide component being modified?

File: src/platform/site-wide/loading-button/LoadingButton.jsx (hunk 7 above)
Comment (vbahinwillit) [reply src/platform/site-wide/loading-button/LoadingButton.jsx:37]: Modification was suggested by Trevor for a 508 defect.
---
Our VA 508 office contact noticed a situation on the /choose-eye-care page where after pressing the Continue button, our gray spinner button announced itself as "unlabeled zero button unavailable" to JAWS + IE11. I've attempted to re-create it by clearing cache and throttling to slow 3G, but wasn't able to reproduce it. Looking at our code for the platform LoadingButton.jsx I have a proposed code solution. Screen shots attached below.

**Review by kelsonic — APPROVED** [review]
Nice work! 🎉 This looks good to me but I would just validate that it looks good in a few of the various apps and that the behavior hasn't changed.

File: src/platform/site-wide/loading-button/LoadingButton.jsx [hunk 8]
```diff
@@ -2,16 +2,40 @@ import React from 'react';
...
+        <span className="sr-only" role="status">
+          {loadingText}
+        </span>
+      </>
+    ) : (
+      children
+    );
+  } else {
+    contents = isLoading ? (
+      <i
```
[the remark this answers is not in the document]
Comment (kelsonic) [reply src/platform/site-wide/loading-button/LoadingButton.jsx:13]: @ncksllvn should have a solution for this 😉

**Pull request merged**